cleanlab/rank.py

```diff
@@ -652,3 +653,106 @@ def get_outlier_scores(
+) -> Union[np.ndarray, Tuple[np.ndarray, NearestNeighbors]]:
+    """Returns an OOD score for each example based on it pred_prob values. Scores lie in [0,1] with smaller values indicating examples
+    that are less typical under the dataset distribution (values near 0 indicate outliers).
+    TODO: what is the score based on.
+
+    Parameters
+    ----------
+    pred_probs : np.ndarray
+      TODO
+
+    confident_thresholds : np.ndarray, default = None
+      TODO
+    labels : np.ndarray, default = None
+      TODO
+
+    adjusted_pred_probs : bool, default = False
+      TODO
+
+    scoring_method : bool, default = False
+      TODO
+
+    t : int, default=1
+      Optional hyperparameter only for advanced users.
+      Controls transformation of distances between examples into similarity scores that lie in [0,1].
+      The transformation applied to distances `x` is `exp(-x*t)`.
+      If you find your scores are all too close to 1, consider increasing `t`,
+      although the relative scores of examples will still have the same ranking across the dataset.
+
+    return_thresholds : bool, default = False
+      Whether the `confident_thresholds` array should also be returned.
+      If True, this function returns a tuple `(ood_scores, confident_thresholds)`.
+    Returns
+    -------
+    ood_scores : np.ndarray
+      Score for each example that roughly corresponds to the likelihood this example stems from the same distribution as
+      the dataset features (i.e. is not an outlier).
+      If ``return_thresholds = True``, then a tuple is returned
+      whose first element is array of `ood_scores` and second is an np.ndarray of `confident_thresholds`.
+    """
+
+    valid_methods = [
+        "entropy",
+        "self_confidence",
+    ]
+
+    # Might be a better place to put this warning
+    if (confident_thresholds is not None or labels is not None) and not adjusted_pred_probs:
+        warnings.warn(
+            f"OOD scores are not adjusted with confident thresholds. If scores need to be adjusted set adjusted_pred_probs = True. Otherwise passing in confident_thresholds and/or labels does not change score calculation.",
+            UserWarning,
+        )
+
+    # TODO warn about both thresholds and labels passed in
+
+    if confident_thresholds is None:
+        if adjusted_pred_probs:
+            if labels is None:
+                raise ValueError(
+                    f"Cannot calculate adjusted_pred_probs without labels. Either pass in labels parameter or set adjusted_pred_probs = False."
+                )
+            confident_thresholds = get_confident_thresholds(
+                labels, pred_probs, multi_label=False
+            )  # should we support multi_label?
+        else:
+            confident_thresholds = 0
+
+    if adjusted_pred_probs:
+        pred_probs = pred_probs - confident_thresholds
+        pred_probs += confident_thresholds.max()
+        pred_probs /= pred_probs.sum(axis=1)[:, None]
```

Comment:
why not just use our existing: `_subtract_confident_thresholds()` function?

https://github.com/cleanlab/cleanlab/blob/master/cleanlab/internal/label_quality_utils.py#L23

Reply:
`_subtract_confident_thresholds() `calculates `confident_thresholds` intrenally. This would not work for some cases here such as user passes in `confident_thresholds` and no `labels`.

Reply:
I see.  But I still think you could alter the `_subtract_confident_thresholds()` function to keep it working for the settings where it's currently being utilized (no change) while also working for this use-case here, right?  

Eg. could call it here with some dummy `labels = None` if there are no labels, and add an optional arg `confident_thresholds` (default value = None) in which case the _subtract...() function only computes the thresholds when this is None, otherwise uses the given thresholds.

Reply:
Maybe I'm missing something and it's more work than is worthwhile to utilize _subtract_confident_thresholds here, but generally will want to minimize code duplication as much as possible!

Reply:
Ok I can definitely open a PR doing that but without any alteration  using `_subtract_confident_thresholds()` won't work

Reply:
Update: Added in newest commit. Not sure if setting pred_probs to optionally = None (to preserve variable ordering for other functions) was the cleanest way of doing that. Open to better suggestions!